azure/services/privatedns/privatedns.go

```diff
@@ -20,202 +20,250 @@ import (
-				set.RecordSetProperties.ARecords = &[]privatedns.ARecord{{
-					Ipv4Address: &record.IP,
-				}}
-			} else if recordType == privatedns.AAAA {
-				set.RecordSetProperties.AaaaRecords = &[]privatedns.AaaaRecord{{
-					Ipv6Address: &record.IP,
-				}}
-			}
-			err := s.client.CreateOrUpdateRecordSet(ctx, s.Scope.ResourceGroup(), zoneSpec.ZoneName, recordType, record.Hostname, set)
-			if err != nil {
-				return errors.Wrapf(err, "failed to create record %s in private DNS zone %s", record.Hostname, zoneSpec.ZoneName)
+	}
+
+	return resErr
+}
+
+func (s *Service) reconcileRecords(ctx context.Context, records []azure.ResourceSpecGetter) error {
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "privatedns.Service.reconcileRecords")
+	defer done()
+
+	var resErr error
+
+	// We go through the list of links to reconcile each one, independently of the result of the previous one.
+	// If multiple errors occur, we return the most pressing one.
+	// Order of precedence (highest -> lowest) is: error that is not an operationNotDoneError (i.e. error creating) -> operationNotDoneError (i.e. creating in progress) -> no error (i.e. created)
+	for _, recordSpec := range records {
+		if _, err := s.recordReconciler.CreateResource(ctx, recordSpec, serviceName); err != nil {
+			if !azure.IsOperationNotDoneError(err) || resErr == nil {
+				resErr = err
 			}
-			log.V(2).Info("successfully created record set", "private dns zone", zoneSpec.ZoneName, "record", record.Hostname)
 		}
 	}
-	return nil
+
+	return resErr
 }
 
 // Delete deletes the private zone and vnet links.
 func (s *Service) Delete(ctx context.Context) error {
-	ctx, log, done := tele.StartSpanWithLogger(ctx, "privatedns.Service.Delete")
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "privatedns.Service.Delete")
 	defer done()
 
-	zoneSpec := s.Scope.PrivateDNSSpec()
-	if zoneSpec != nil {
-		for _, linkSpec := range zoneSpec.Links {
-			// If the virtual network link is not managed by capz, skip its removal
-			isVnetLinkManaged, err := s.isVnetLinkManaged(ctx, s.Scope.ResourceGroup(), zoneSpec.ZoneName, linkSpec.LinkName)
-			if err != nil && !azure.ResourceNotFound(err) {
-				return errors.Wrapf(err, "could not get vnet link state of %s in resource group %s", zoneSpec.ZoneName, s.Scope.ResourceGroup())
-			}
-			if !isVnetLinkManaged {
-				log.V(2).Info("Skipping vnet link deletion for unmanaged vnet link", "vnet link", linkSpec.LinkName, "private dns zone", zoneSpec.ZoneName)
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
+
+	zoneSpec, links, _ := s.Scope.PrivateDNSSpec()
+	if zoneSpec == nil {
+		return nil
+	}
+
+	if err := s.deleteLinks(ctx, links); err != nil {
+		s.Scope.UpdateDeleteStatus(infrav1.PrivateDNSReadyCondition, serviceName, err)
+		return err
+	}
+
+	if err := s.deleteZone(ctx, zoneSpec); err != nil {
+		s.Scope.UpdateDeleteStatus(infrav1.PrivateDNSReadyCondition, serviceName, err)
```

Comment:
Same for skipping delete on an unmanaged zone. So I guess tl;dr is that if we go through the reconciler or deletion and see that we don't have any managed specs at all, then we don't update the status. @CecileRobertMichon Does this approach make sense?